File: Java/First-Missing-Positive.java
public class Solution {
    /**
     * @param A: an array of integers
     * @return: an integer
     */
    public int firstMissingPositive(int[] A) {
        // write your code here
        for (int i = 0; i < A.length;) {
            int temp = A[i];
            if (temp > 0 && temp <= A.length && A[temp - 1] != temp) {
                int cur = A[temp - 1];
                A[temp - 1] = temp;
                A[i] = cur;
            } else {
                i ++;
            }
        }
        for (int i = 0; i < A.length; i ++) {
            if (i + 1 != A[i]) return i + 1;
        }
        return A.length + 1;
    }
}

File: Python/First-Missing-Positive.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        # write your code here
        N = len(A)
        for i in xrange(N):
            while A[i] > 0 and A[i] <= N and A[i] != i + 1:
                if A[A[i] - 1] == A[i]:
                    break
                temp = A[A[i] - 1]
                A[A[i] - 1] = A[i]
                A[i] = temp
        for i in xrange(N):
            if A[i] != i + 1:
                return i + 1
        return N + 1